lib/state/backend/authenticator.dart

```diff
@@ -0,0 +1,130 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_facebook_auth/flutter_facebook_auth.dart';
+import 'package:google_sign_in/google_sign_in.dart';
+import 'package:runningapp/database/repository.dart';
+import 'package:runningapp/state/auth/constants/constants.dart';
+import 'package:runningapp/state/auth/models/auth_results.dart';
+import 'package:runningapp/state/auth/typedefs.dart';
+
+class Authenticator {
+  // Getters
+  Stream<User?> get authStateChanges =>
+      FirebaseAuth.instance.authStateChanges();
+  UserId? get userId => FirebaseAuth.instance.currentUser?.uid;
+  bool get isAlreadyLoggedIn => userId != null;
+  String get displayName =>
+      FirebaseAuth.instance.currentUser?.displayName ?? '';
+  String? get email => FirebaseAuth.instance.currentUser?.email;
+
+  // Logout
+  Future<void> logOut() async {
+    await FirebaseAuth.instance.signOut();
+    await GoogleSignIn().signOut();
+    FirebaseFirestore.instance.terminate();
+    FirebaseFirestore.instance.clearPersistence();
+    // await FacebookAuth.instance.logOut(); TODO next time
+  }
+
+  Future<AuthResult> loginWithFacebook() async {
+    final loginResult = await FacebookAuth.instance.login();
+    final token = loginResult.accessToken?.token;
+    if (token == null) {
+      return AuthResult.aborted;
+    }
+    final oauthCredentials = FacebookAuthProvider.credential(token);
+
+    try {
+      await FirebaseAuth.instance.signInWithCredential(oauthCredentials);
+      return AuthResult.success;
+    } on FirebaseAuthException catch (e) {
+      final email = e.email;
+      final credential = e.credential;
+      if (e.code == Constants.accountExistsWithDifferentCredentialsError &&
+          email != null &&
+          credential != null) {
+        final providers =
+            await FirebaseAuth.instance.fetchSignInMethodsForEmail(email);
```

Comment:
Just a note for this, this deprecated due to security reasons but Firebase themsleves have not provided an offical replacement method. In the future, the general consensus is to try to avoid relying on this method if possible

[fetchSignInMethodsForEmail](https://github.com/firebase/firebase-android-sdk/issues/5586)